From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Maciej Patelczyk <maciej.patelczyk@intel.com>
Date: Tue, 13 Jan 2026 14:06:12 +0100
Subject: fixup! drm/xe/eudebug: Prelim rework for Xe EU debug

Eudebug prelim fix for exec queue property.

Detach prelim flags/property from non-prelim ones.
Provides 2 separate namespaces for them.

Signed-off-by: Maciej Patelczyk <maciej.patelczyk@intel.com>
(backported from commit 81a0843b1b211787d4c2fb279815521be9ba08ac drm-tip-eudebug)
Signed-off-by: Pravalika Gurram <pravalika.gurram@intel.com>
---
 drivers/gpu/drm/xe/xe_exec_queue.c | 70 +++++++++++++++++++++++++-----
 1 file changed, 59 insertions(+), 11 deletions(-)

diff --git a/drivers/gpu/drm/xe/xe_exec_queue.c b/drivers/gpu/drm/xe/xe_exec_queue.c
index d63a6cebe..3d7df8f8a 100644
--- a/drivers/gpu/drm/xe/xe_exec_queue.c
+++ b/drivers/gpu/drm/xe/xe_exec_queue.c
@@ -1063,7 +1063,14 @@ static const xe_exec_queue_set_property_fn exec_queue_set_property_funcs[] = {
 							exec_queue_set_multi_queue_priority,
 	[DRM_XE_EXEC_QUEUE_SET_DISABLE_STATE_CACHE_PERF_FIX] =
 							exec_queue_set_state_cache_perf_fix,
-	[PRELIM_DRM_XE_EXEC_QUEUE_SET_PROPERTY_EUDEBUG] = exec_queue_set_eudebug,
+};
+
+static const xe_exec_queue_set_property_fn prelim_exec_queue_set_property_funcs[] = {
+#define MAKE_TABLE_IDX(id)[PRELIM_DRM_XE_EXEC_QUEUE_SET_PROPERTY_##id & PRELIM_DRM_XE_VALUE_MASK]
+
+	MAKE_TABLE_IDX(EUDEBUG) = exec_queue_set_eudebug,
+
+#undef MAKE_TABLE_IDX
 };
 
 /**
@@ -1098,9 +1105,32 @@ int xe_exec_queue_set_property_ioctl(struct drm_device *dev, void *data,
 	if (XE_IOCTL_DBG(xe, !q))
 		return -ENOENT;
 
-	idx = array_index_nospec(args->property,
-				 ARRAY_SIZE(exec_queue_set_property_funcs));
-	ret = exec_queue_set_property_funcs[idx](xe, q, args->value);
+	if (args->property & PRELIM_DRM_XE_VALUE_INDICATOR) {
+		if (XE_IOCTL_DBG(xe, PRELIM_DRM_XE_VALUE(args->property) >=
+				 ARRAY_SIZE(prelim_exec_queue_set_property_funcs)))
+			return -EINVAL;
+
+		idx = array_index_nospec(PRELIM_DRM_XE_VALUE(args->property),
+					 ARRAY_SIZE(prelim_exec_queue_set_property_funcs));
+
+		if (!prelim_exec_queue_set_property_funcs[idx])
+			return -EINVAL;
+
+		ret = prelim_exec_queue_set_property_funcs[idx](xe, q, args->value);
+	} else {
+		if (XE_IOCTL_DBG(xe, args->property >=
+				 ARRAY_SIZE(exec_queue_set_property_funcs)))
+			return -EINVAL;
+
+		idx = array_index_nospec(args->property,
+					 ARRAY_SIZE(exec_queue_set_property_funcs));
+
+		if (!exec_queue_set_property_funcs[idx])
+			return -EINVAL;
+
+		ret = exec_queue_set_property_funcs[idx](xe, q, args->value);
+	}
+
 	if (XE_IOCTL_DBG(xe, ret))
 		goto err_post_lookup;
 
@@ -1151,9 +1181,7 @@ static int exec_queue_user_ext_set_property(struct xe_device *xe,
 	if (XE_IOCTL_DBG(xe, err))
 		return -EFAULT;
 
-	if (XE_IOCTL_DBG(xe, ext.property >=
-			 ARRAY_SIZE(exec_queue_set_property_funcs)) ||
-	    XE_IOCTL_DBG(xe, ext.pad) ||
+	if (XE_IOCTL_DBG(xe, ext.pad) ||
 	    XE_IOCTL_DBG(xe, ext.property != DRM_XE_EXEC_QUEUE_SET_PROPERTY_PRIORITY &&
 			 ext.property != DRM_XE_EXEC_QUEUE_SET_PROPERTY_TIMESLICE &&
 			 ext.property != DRM_XE_EXEC_QUEUE_SET_PROPERTY_PXP_TYPE &&
@@ -1164,16 +1192,36 @@ static int exec_queue_user_ext_set_property(struct xe_device *xe,
 			 ext.property != PRELIM_DRM_XE_EXEC_QUEUE_SET_PROPERTY_EUDEBUG))
 		return -EINVAL;
 
-	idx = array_index_nospec(ext.property, ARRAY_SIZE(exec_queue_set_property_funcs));
-	if (!exec_queue_set_property_funcs[idx])
-		return -EINVAL;
+	if (ext.property &  PRELIM_DRM_XE_VALUE_INDICATOR) {
+		if (XE_IOCTL_DBG(xe, PRELIM_DRM_XE_VALUE(ext.property) >=
+				 ARRAY_SIZE(prelim_exec_queue_set_property_funcs)))
+			return -EINVAL;
+
+		idx = array_index_nospec(PRELIM_DRM_XE_VALUE(ext.property),
+					 ARRAY_SIZE(prelim_exec_queue_set_property_funcs));
+		if (!prelim_exec_queue_set_property_funcs[idx])
+			return -EINVAL;
+	} else {
+		if (XE_IOCTL_DBG(xe, ext.property >=
+				 ARRAY_SIZE(exec_queue_set_property_funcs)))
+			return -EINVAL;
+
+		idx = array_index_nospec(ext.property, ARRAY_SIZE(exec_queue_set_property_funcs));
+		if (!exec_queue_set_property_funcs[idx])
+			return -EINVAL;
+	}
 
 	*properties |= BIT_ULL(idx);
 	err = exec_queue_user_ext_check(q, *properties);
 	if (XE_IOCTL_DBG(xe, err))
 		return err;
 
-	return exec_queue_set_property_funcs[idx](xe, q, ext.value);
+	if (ext.property &  PRELIM_DRM_XE_VALUE_INDICATOR)
+		err =  prelim_exec_queue_set_property_funcs[idx](xe, q, ext.value);
+	else
+		err =  exec_queue_set_property_funcs[idx](xe, q, ext.value);
+
+	return err;
 }
 
 typedef int (*xe_exec_queue_user_extension_fn)(struct xe_device *xe,
-- 
2.43.0

